### PR TITLE
Add support for OCR2 Key Management - No creation for now 

### PR DESCRIPTION
**Pull request by hans-kessock**
## Description

`You may now view OCR2 keys on the same Key Management UI as the other keys, following the same structures and patterns as the original OCR keys (except with different data of course)`

`Please note that OCR2 key creation is not supported at this time as OCR2 keys require the specification of a chain family and that list is now, as far as any client is concerned, dynamically generated by the core node - this usually mean a simple query, but with the very soon to be added support for blockchain plugins, I am very leery of exposing this from the node just yet`

`There is also an addition to resolve the unhandled session timeout issue (exhibited by the Key Management page) that stems from that screen/view not having a query that would trigger session timeout at the appropriate level`

## Steps to Test

1. `yarn && yarn setup`
2. `yarn start`
3. ...etc

- [ x] This PR has an accompanying changeset if needed.

### CODE REVIEW

**Comment by hans-kessock** [comment]
Tests are breaking now because of the resolver code being only aware of EVM and SOLANA (as well as being case sensitive which is problematic because they are stored by the keystore in lowercase I believe.)

I can update the resolver to be aware of starknet, but that is just kicking the can down the road because the core is moving away from hardcoded knowledge about chain families (which is why the OCR2 key management support I'm adding does not yet support creation of OCR2 keys.)

We *might* want to remove chain family name validation from the integration test until this is resolved architecturally on the core node side; although, it will likely no longer be a part  of the graphql generated definitions going forward - but that's not certain yet.)

@jkongie / @HenryNguyen5 - what do y'all think?  I'm a newbie here at Chainlink, so feel free to club away ;)

File: support/factories/gql/fetchOCR2KeyBundles.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+// buildOCR2KeyBundle builds a ocr2 key bundle for the FetchOCR2KeyBundles query.
+export function buildOCR2KeyBundle(
+  overrides?: Partial<Ocr2KeyBundlesPayload_ResultsFields>,
+): Ocr2KeyBundlesPayload_ResultsFields {
+  return {
+    __typename: 'OCR2KeyBundle',
+    id: '68ae4225aa9fd932e62a2411e4c757a9fa72de45426ac455801a6f08cb4392b3',
+    chainType: 'evm',
```
Comment (jkongie) [inline support/factories/gql/fetchOCR2KeyBundles.ts:8]: You should change this to `EVM` to match the correct types

**Comment by jkongie** [comment]
I believe that you may to update the GQL schema with the new type anyway due to how the typecheck works, just to get this in. Since you are using `starknet` in the tests as a value, you will need to ensure that the frontend is compatible with the backend enum declaration.

**Comment by hans-kessock** [comment]
> 

@jkongie - that's the weirdness though.  The backend enum is both hard coded (currently) and all caps.  The OCR2 key family is created/carried/persisted lowercase.

I'll double check that the query doesn't perform any mapping/transformations when bundling the key data.  If it does, I'll update the enumeration - otherwise it is indicative that the enumeration names and the keybundle values have been disconnected all along for OCR2 keys.

The OCR2 client side graphql bundling code already handles 'starknet' and other lowercase names just fine, it's only breaking in the CI tests.

I can modify my tests to avoid names it isn't aware of - since the real world usage doesn't use the forms the tests expect anyhow - but I may change the enumeration instead if it doesn't break other things.

File: support/factories/gql/fetchOCR2KeyBundles.ts (hunk 1 above)
Comment (hans-kessock) [reply support/factories/gql/fetchOCR2KeyBundles.ts:8]: Updating the graphql enums to remove terra and add starknet, the resolver, and will update the fetcher as well.  Thanks!

Comment (hans-kessock) [reply support/factories/gql/fetchOCR2KeyBundles.ts]: Fixed, and the second key in the bundle now matches the new graphQL side (in the process of being merged in core) of 'STARKNET'

**Comment by HenryNguyen5** [comment]
Hey Hans,
You could create a useQueryErrorHandler hook that is a copy/paste of the useMutationErrorHandler, then, you can either:

Use the onError callback handler of useOCR2KeysQuery to sign out and redirect on UNAUTHORIZED apollo errors like so:
```ts
const handleQueryErrors = useQueryErrorHandler()
const { data, loading, refetch } = useOCR2KeysQuery({
    fetchPolicy: 'network-only',
    onError: handleQueryErrors
  })
```
or

Call the handler directly
```ts
const handleQueryErrors = useQueryErrorHandler()
const { data, loading, refetch, error } = useOCR2KeysQuery({
    fetchPolicy: 'network-only',
    onError: handleQueryErrors
  })
handleQueryErrors(error)
````
Note that we make our own handler since the error handling implementation of queries vs mutations may diverge over time.
You will notice that we will have 3 error handling hooks afterwards:

- useErrorHandler
- useMutationErrorHandler
- useQueryErrorHandler

It may be wise to refactor the 3 hooks to extract common functionality, at a glance, they do the same logic when handling unauth errors.

If i understand you correctly, this will give you the intended result of notifying users on query error within the parent container component without having the child view component do that handling (edited)

File: support/factories/gql/fetchOCR2KeyBundles.ts (hunk 1 above)
Comment (george-dorin) [reply support/factories/gql/fetchOCR2KeyBundles.ts]: It seems like starknet is not going to be merged any time soon, skipping it for now

**Comment by george-dorin** [comment]
> Hey Hans, You could create a useQueryErrorHandler hook that is a copy/paste of the useMutationErrorHandler, then, you can either:
> 
> Use the onError callback handler of useOCR2KeysQuery to sign out and redirect on UNAUTHORIZED apollo errors like so:
> 
> ```ts
> const handleQueryErrors = useQueryErrorHandler()
> const { data, loading, refetch } = useOCR2KeysQuery({
>     fetchPolicy: 'network-only',
>     onError: handleQueryErrors
>   })
> ```
> 
>  
> or
> 
> Call the handler directly
> 
> ```ts
> const handleQueryErrors = useQueryErrorHandler()
> const { data, loading, refetch, error } = useOCR2KeysQuery({
>     fetchPolicy: 'network-only',
>     onError: handleQueryErrors
>   })
> handleQueryErrors(error)
> ```
> 
>  
> Note that we make our own handler since the error handling implementation of queries vs mutations may diverge over time. You will notice that we will have 3 error handling hooks afterwards:
> 
> * useErrorHandler
> * useMutationErrorHandler
> * useQueryErrorHandler
> 
> It may be wise to refactor the 3 hooks to extract common functionality, at a glance, they do the same logic when handling unauth errors.
> 
> You could create a useQueryErrorHandler hook that is a copy/paste of the useMutationErrorHandler, then, you can either

Added in [`0e400df` (#26)](https://github.com/smartcontractkit/operator-ui/pull/26/commits/0e400df5124762c81b45cdd040a06268f88c7763)

**Comment by george-dorin** [comment]
As far as I know, starknet support for GQL is not going to be implemented any time soon so we should ignore it for now

File: src/hooks/useQueryErrorHandler.tsx [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+          return
+        }
+      })
+
+      dispatch(notifyErrorMsg(error.message))
+    } else if (error instanceof Error) {
+      dispatch(notifyErrorMsg(error.message))
+    } else {
+      dispatch(notifyErrorMsg('An error occured'))
+    }
```
Comment (HenryNguyen5) [inline src/hooks/useQueryErrorHandler.tsx:38]: Could just do `dispatch(notifyErrorMsg(error.?message || 'An error occured'))`

File: src/hooks/useQueryErrorHandler.tsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+// query operation. If the error returned is an authentication error, it
+// signs the user out and redirects them to the sign in page, otherwise it
+// displays an alert with the error message.
+export const useQueryErrorHandler = () => {
+  const [error, setError] = React.useState<unknown>()
+  const history = useHistory()
+  const dispatch = useDispatch()
+
+  React.useEffect(() => {
+    if (error === null || error === undefined) {
```
Comment (HenryNguyen5) [inline src/hooks/useQueryErrorHandler.tsx:19]: `if(!error) `

File: src/hooks/useQueryErrorHandler.tsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+import { notifyErrorMsg } from 'actionCreators'
+import { ApolloError } from '@apollo/client'
+import { receiveSignoutSuccess } from 'actionCreators'
+
+// useQueryErrorHandler handles an unknown error which is caught from a
+// query operation. If the error returned is an authentication error, it
+// signs the user out and redirects them to the sign in page, otherwise it
+// displays an alert with the error message.
+export const useQueryErrorHandler = () => {
```
Comment (HenryNguyen5) [inline src/hooks/useQueryErrorHandler.tsx:13]: Prefer TSDoc format here for better editor hinting

File: src/hooks/useQueryErrorHandler.tsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+      dispatch(notifyErrorMsg(error.message))
+    } else if (error instanceof Error) {
+      dispatch(notifyErrorMsg(error.message))
+    } else {
+      dispatch(notifyErrorMsg('An error occured'))
+    }
+  }, [dispatch, error, history])
+
+  return { handleQueryError: setError }
```
Comment (HenryNguyen5) [inline src/hooks/useQueryErrorHandler.tsx:41]: Might as well have the name be `handleQueryError` for one less level of indirection

File: src/screens/KeyManagement/OCR2Keys.tsx [hunk 6]
```diff
@@ -55,7 +54,7 @@ export const OCR2Keys = () => {
           break
       }
     } catch (e) {
-      handleMutationError(e)
+      handleQueryError(e)
```
Comment (HenryNguyen5) [inline src/screens/KeyManagement/OCR2Keys.tsx:57]: Shouldn't this still be a mutation error? We're deleting resources

**Review by HenryNguyen5 — CHANGES_REQUESTED** [review]

File: src/screens/KeyManagement/OCR2Keys.tsx [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    } catch (e) {
+      handleQueryError(e)
+    }
+  }
+
+  return (
+    <OCR2KeysCard
+      loading={loading}
+      data={data}
+      errorMsg={error?.message}
```
Comment (HenryNguyen5) [inline src/screens/KeyManagement/OCR2Keys.tsx:65]: Do we still need this prop if we have an error handler already set?

**Review by HenryNguyen5 — CHANGES_REQUESTED** [review]

File: src/hooks/useQueryErrorHandler.tsx [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  const dispatch = useDispatch()
+
+  React.useEffect(() => {
+    if (error === null || error === undefined) {
+      return
+    } else if (error instanceof ApolloError) {
+      // Check for an authentication error and logout
+      error.graphQLErrors.forEach((err) => {
+        if (err.extensions?.code == 'UNAUTHORIZED') {
+          dispatch(receiveSignoutSuccess())
```
[the remark this answers is not in the document]
Comment (george-dorin) [reply src/hooks/useQueryErrorHandler.tsx:27]: Changed in [`2079e2a` (#26)](https://github.com/smartcontractkit/operator-ui/pull/26/commits/2079e2a0f719b07775cdfe89764f91e91e9ea3aa)

File: src/hooks/useQueryErrorHandler.tsx (hunk 3 above)
Comment (george-dorin) [reply src/hooks/useQueryErrorHandler.tsx]: Changed in [`2079e2a` (#26)](https://github.com/smartcontractkit/operator-ui/pull/26/commits/2079e2a0f719b07775cdfe89764f91e91e9ea3aa)

File: src/hooks/useQueryErrorHandler.tsx (hunk 2 above)
Comment (george-dorin) [reply src/hooks/useQueryErrorHandler.tsx]: Changed in [`2079e2a` (#26)](https://github.com/smartcontractkit/operator-ui/pull/26/commits/2079e2a0f719b07775cdfe89764f91e91e9ea3aa)

File: src/hooks/useQueryErrorHandler.tsx (hunk 5 above)
Comment (george-dorin) [reply src/hooks/useQueryErrorHandler.tsx]: Changed in [`2079e2a` (#26)](https://github.com/smartcontractkit/operator-ui/pull/26/commits/2079e2a0f719b07775cdfe89764f91e91e9ea3aa)

File: src/screens/KeyManagement/OCR2Keys.tsx (hunk 7 above)
Comment (george-dorin) [reply src/screens/KeyManagement/OCR2Keys.tsx]: Removed in [`2079e2a` (#26)](https://github.com/smartcontractkit/operator-ui/pull/26/commits/2079e2a0f719b07775cdfe89764f91e91e9ea3aa)

**Pull request merged**